In [181]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

<h3>Wczytaj końcowy i przetworzony zbiór danych z zajęć Lab2.<h3>

In [182]:
df = pd.read_csv('titanic_lab2.csv')
df.head(20)

Unnamed: 0  pclass  survived  \
0            0       1         1   
1            1       1         1   
2            2       1         0   
3            3       1         0   
4            4       1         0   
5            5       1         1   
6            6       1         1   
7            7       1         0   
8            8       1         1   
9            9       1         0   
10          10       1         0   
11          11       1         1   
12          12       1         1   
13          13       1         1   
14          14       1         1   
15          15       1         0   
16          16       1         0   
17          17       1         1   
18          18       1         1   
19          19       1         0   

                                                 name     sex      age  sibsp  \
0                       Allen, Miss. Elisabeth Walton  female  29.0000      0   
1                      Allison, Master. Hudson Trevor    male   0.9167      1   
2                        Allison, Miss. Helen Loraine  female   2.0000      1   
3                Allison, Mr. Hudson Joshua Creighton    male  30.0000      1   
4     Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female  25.0000      1   
5                                 Anderson, Mr. Harry    male  48.0000      0   
6                   Andrews, Miss. Kornelia Theodosia  female  63.0000      1   
7                              Andrews, Mr. Thomas Jr    male  39.0000      0   
8       Appleton, Mrs. Edward Dale (Charlotte Lamson)  female  53.0000      2   
9                             Artagaveytia, Mr. Ramon    male  71.0000      0   
10                             Astor, Col. John Jacob    male  47.0000      1   
11  Astor, Mrs. John Jacob (Madeleine Talmadge Force)  female  18.0000      1   
12                      Aubart, Mme. Leontine Pauline  female  24.0000      0   
13                       Barber, Miss. Ellen 'Nellie'  female  26.0000      0   
14               Barkworth, Mr. Algernon Henry Wilson    male  80.0000      0   
15                                Baumann, Mr. John D    male      NaN      0   
16                           Baxter, Mr. Quigg Edmond    male  24.0000      0   
17    Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female  50.0000      0   
18                              Bazzani, Miss. Albina  female  32.0000      0   
19                               Beattie, Mr. Thomson    male  36.0000      0   

    parch    ticket      fare  ... embarked boat   body  \
0       0     24160  211.3375  ...        S    2    NaN   
1       2    113781  151.5500  ...        S   11    NaN   
2       2    113781  151.5500  ...        S  NaN    NaN   
3       2    113781  151.5500  ...        S  NaN  135.0   
4       2    113781  151.5500  ...        S  NaN    NaN   
5       0     19952   26.5500  ...        S    3    NaN   
6       0     13502   77.9583  ...        S   10    NaN   
7       0    112050    0.0000  ...        S  NaN    NaN   
8       0     11769   51.4792  ...        S    D    NaN   
9       0  PC 17609   49.5042  ...        C  NaN   22.0   
10      0  PC 17757  227.5250  ...        C  NaN  124.0   
11      0  PC 17757  227.5250  ...        C    4    NaN   
12      0  PC 17477   69.3000  ...        C    9    NaN   
13      0     19877   78.8500  ...        S    6    NaN   
14      0     27042   30.0000  ...        S    B    NaN   
15      0  PC 17318   25.9250  ...        S  NaN    NaN   
16      1  PC 17558  247.5208  ...        C  NaN    NaN   
17      1  PC 17558  247.5208  ...        C    6    NaN   
18      0     11813   76.2917  ...        C    8    NaN   
19      0     13050   75.2417  ...        C    A    NaN   

                          home_dest KolumnaNullCabin  KolumnaNullBoat  \
0                      St Louis, MO                0                0   
1   Montreal, PQ / Chesterville, ON                0                0   
2   Montreal, PQ / Chesterville, ON                0                1   
3   Montreal,

<h3>Zapoznaj się z funkcja train_test_split wchodzącą w skład biblioteki
Scikit-learn. Zapisz swoje obserwacje. <h3>

<h6>Funckja ta pozwala podzielić w losowy sposób zbiór danych na zbiór treningowy i testowy.

<h3>Stwórz zmienną do której zapiszesz listę z nazwami trzech kolumn –
kabiny, zredukowane kabiny oraz płeć. Nazwij ją col_name<h3>

In [183]:
col_name = ['cabin', 'cabinReduced', 'sex']

<h3>Podziel zbiór na treningowy i testowy używając train_test_split.
Jako zmienną niezależną ustaw dane składające się z kolumn o nazwach
zapisanych w col_name. Jako zmienną zależną ustaw kolumnę mówiącą o
tym czy ktoś przeżył czy nie. Ustaw rozmiar zbioru testowego na 20 lub
30% całości. Parametr random_state ustaw jako 0.
Wyświetl wymiary zbiorów X_train, X_test, y_train, y_test używając
.shape() i skomentuj wyniki.<h3>

In [184]:
X_train, X_test, y_train, y_test = train_test_split(df[col_name],df['survived'], test_size=0.30, random_state=0)

In [185]:
np.shape(X_train)

(916, 3)

In [186]:
np.shape(X_test)

(393, 3)

In [187]:
np.shape(y_train)

(916,)

In [188]:
np.shape(y_test)

(393,)

In [189]:
393/(393+916)*100


30.02291825821238

<h6> Powyższa funkcja podzieliła zbiór pierwotny na cztery listy: X_train, X_test, y_train, y_test, w stosunku 30% do 70% - zgodnie z podanym parametrem<h6>

<h3>Zbadaj w zależności od kardynalności danej zmiennej czy rozkład etykiet
dla poszczególnych cech w zbiorach testowych i treningowych jest
równomierny. <h3>

In [190]:
print(X_test)

     cabin cabinReduced     sex
1139   NaN            n    male
533    NaN            n  female
459    NaN            n    male
1150   NaN            n    male
393    NaN            n    male
...    ...          ...     ...
914    NaN            n    male
580    NaN            n  female
1080   NaN            n    male
1249   F38            F    male
88     NaN            n  female

[393 rows x 3 columns]


In [191]:
unique_test = [x for x in X_test.cabin.unique() if x not in X_train.cabin.unique()]
len(unique_test)

37

In [192]:
unique_test = [x for x in X_test.cabinReduced.unique() if x not in X_train.cabinReduced.unique()]
len(unique_test)

0

In [193]:
unique_test = [x for x in X_test.sex.unique() if x not in X_train.sex.unique()]
len(unique_test)

0

<h6> Dla małej kardynalności zmiennych rozkład etykiet jest równomierny - zarówno etykiety z cabinReduced jak i z sex ze zbioru testowego występują w zbiorze treningowym. Dla zmiennej o dużej kardynalności (cabin) występują wartości występujące w zbiorze testowym a niewystępujące w zbiorze treningowym.

<h3>Wykonaj kodowanie zmiennych kategorycznych do zmiennych
liczbowych. Wykorzystaj pętlę for i metodę enumerate(). Wynik powinien
być podobny do tego na obrazku poniżej (UWAGA! Na rysunku jest
tylko część rozwiązania i przykład tylko dla zmiennej kabina).<h3>

In [194]:
cabin_dict={}
for i, value in enumerate(df.cabin.unique()):
    cabin_dict[value]=i
cabin_dict

{'B5': 0,
 'C22 C26': 1,
 'E12': 2,
 'D7': 3,
 'A36': 4,
 'C101': 5,
 nan: 6,
 'C62 C64': 7,
 'B35': 8,
 'A23': 9,
 'B58 B60': 10,
 'D15': 11,
 'C6': 12,
 'D35': 13,
 'C148': 14,
 'C97': 15,
 'B49': 16,
 'C99': 17,
 'C52': 18,
 'T': 19,
 'A31': 20,
 'C7': 21,
 'C103': 22,
 'D22': 23,
 'E33': 24,
 'A21': 25,
 'B10': 26,
 'B4': 27,
 'E40': 28,
 'B38': 29,
 'E24': 30,
 'B51 B53 B55': 31,
 'B96 B98': 32,
 'C46': 33,
 'E31': 34,
 'E8': 35,
 'B61': 36,
 'B77': 37,
 'A9': 38,
 'C89': 39,
 'A14': 40,
 'E58': 41,
 'E49': 42,
 'E52': 43,
 'E45': 44,
 'B22': 45,
 'B26': 46,
 'C85': 47,
 'E17': 48,
 'B71': 49,
 'B20': 50,
 'A34': 51,
 'C86': 52,
 'A16': 53,
 'A20': 54,
 'A18': 55,
 'C54': 56,
 'C45': 57,
 'D20': 58,
 'A29': 59,
 'C95': 60,
 'E25': 61,
 'C111': 62,
 'C23 C25 C27': 63,
 'E36': 64,
 'D34': 65,
 'D40': 66,
 'B39': 67,
 'B41': 68,
 'B102': 69,
 'C123': 70,
 'E63': 71,
 'C130': 72,
 'B86': 73,
 'C92': 74,
 'A5': 75,
 'C51': 76,
 'B42': 77,
 'C91': 78,
 'C125': 79,
 'D10 D12': 80,
 'B82 

In [195]:
cabinReduced_dict={}
for i, value in enumerate(df.cabinReduced.unique()):
    cabinReduced_dict[value]=i
cabinReduced_dict

{'B': 0, 'C': 1, 'E': 2, 'D': 3, 'A': 4, 'n': 5, 'T': 6, 'F': 7, 'G': 8}

In [196]:
sex_dict={}
for i, value in enumerate(df.sex.unique()):
    sex_dict[value]=i
sex_dict

{'female': 0, 'male': 1}

<h3>Zastąp etykiety zmiennej (tu przykład dla kabina) słownikiem
stworzonym w kroku 6. Do tego będzie potrzebne mapowanie.
<h3>

In [197]:
X_train["cabin_map"] = X_train["cabin"].map(cabin_dict)
X_train[["cabin", "cabin_map"]].head(20)

cabin  cabin_map
501       NaN          6
588       NaN          6
402       NaN          6
1193      NaN          6
686       NaN          6
971       NaN          6
117       E36         64
540       NaN          6
294       C68        157
261       E24         30
587       NaN          6
489       NaN          6
2     C22 C26          1
405       NaN          6
1284      NaN          6
338       NaN          6
356       NaN          6
985       NaN          6
182       NaN          6
1027      NaN          6

In [198]:
X_test["cabin_map"] = X_test["cabin"].map(cabin_dict)
X_test[["cabin", "cabin_map"]].head(20)

cabin  cabin_map
1139              NaN          6
533               NaN          6
459               NaN          6
1150              NaN          6
393               NaN          6
1189               G6        185
5                 E12          2
231              C104        130
330               NaN          6
887               NaN          6
531               NaN          6
790               NaN          6
427               NaN          6
1260              NaN          6
251   B57 B59 B63 B66        138
486               NaN          6
436               NaN          6
31                A31         20
186               NaN          6
1051              NaN          6

In [199]:
X_train["cabinReduced_map"] = X_train["cabinReduced"].map(cabinReduced_dict)
X_train[["cabinReduced", "cabinReduced_map"]].head(20)

cabinReduced  cabinReduced_map
501             n                 5
588             n                 5
402             n                 5
1193            n                 5
686             n                 5
971             n                 5
117             E                 2
540             n                 5
294             C                 1
261             E                 2
587             n                 5
489             n                 5
2               C                 1
405             n                 5
1284            n                 5
338             n                 5
356             n                 5
985             n                 5
182             n                 5
1027            n                 5

In [200]:
X_test["cabinReduced_map"] = X_test["cabinReduced"].map(cabinReduced_dict)
X_test[["cabinReduced", "cabinReduced_map"]].head(20)

cabinReduced  cabinReduced_map
1139            n                 5
533             n                 5
459             n                 5
1150            n                 5
393             n                 5
1189            G                 8
5               E                 2
231             C                 1
330             n                 5
887             n                 5
531             n                 5
790             n                 5
427             n                 5
1260            n                 5
251             B                 0
486             n                 5
436             n                 5
31              A                 4
186             n                 5
1051            n                 5

In [201]:
X_train["sex_map"] = X_train["sex"].map(sex_dict)
X_train[["sex", "sex_map"]].head(20)

sex  sex_map
501   female        0
588   female        0
402   female        0
1193    male        1
686   female        0
971     male        1
117   female        0
540   female        0
294     male        1
261     male        1
587     male        1
489   female        0
2     female        0
405     male        1
1284    male        1
338     male        1
356     male        1
985     male        1
182   female        0
1027    male        1

In [202]:
X_test["sex_map"] = X_test["sex"].map(sex_dict)
X_test[["sex", "sex_map"]].head(20)

sex  sex_map
1139    male        1
533   female        0
459     male        1
1150    male        1
393     male        1
1189  female        0
5       male        1
231     male        1
330     male        1
887     male        1
531     male        1
790     male        1
427     male        1
1260  female        0
251   female        0
486     male        1
436   female        0
31      male        1
186   female        0
1051    male        1

<h3>. Sprawdź liczbę brakujących wartości w zmodyfikowanych zbiorach.
Zapisz wyniki i skomentuj. <h3>

In [203]:
X_train.isnull().sum()

cabin               702
cabinReduced          0
sex                   0
cabin_map             0
cabinReduced_map      0
sex_map               0
dtype: int64

In [204]:
X_test.isnull().sum()

cabin               312
cabinReduced          0
sex                   0
cabin_map             0
cabinReduced_map      0
sex_map               0
dtype: int64

In [205]:
len(X_train)
702/916

0.7663755458515283

In [206]:
len(X_test)
312/393

0.7938931297709924

<h6>Dość spore braki danych widać w kategorii cabin. W przypadku zbioru testowego jak i treningowego jest to prawie 80%<h6>

In [207]:
cabin_missing = [x for x in X_test.cabin_map if x==6]
len(cabin_missing)

312

In [208]:
cabinReduced_missing = [x for x in X_test.cabinReduced_map if x==5]
len(cabinReduced_missing)

312

<h3>Zastąp brakujące wartości liczbą 0. Czy jest to najlepsze wyjście?

In [209]:
cabin_dict.update({np.nan:0, 'B5':6})
cabin_dict

{'B5': 6,
 'C22 C26': 1,
 'E12': 2,
 'D7': 3,
 'A36': 4,
 'C101': 5,
 nan: 0,
 'C62 C64': 7,
 'B35': 8,
 'A23': 9,
 'B58 B60': 10,
 'D15': 11,
 'C6': 12,
 'D35': 13,
 'C148': 14,
 'C97': 15,
 'B49': 16,
 'C99': 17,
 'C52': 18,
 'T': 19,
 'A31': 20,
 'C7': 21,
 'C103': 22,
 'D22': 23,
 'E33': 24,
 'A21': 25,
 'B10': 26,
 'B4': 27,
 'E40': 28,
 'B38': 29,
 'E24': 30,
 'B51 B53 B55': 31,
 'B96 B98': 32,
 'C46': 33,
 'E31': 34,
 'E8': 35,
 'B61': 36,
 'B77': 37,
 'A9': 38,
 'C89': 39,
 'A14': 40,
 'E58': 41,
 'E49': 42,
 'E52': 43,
 'E45': 44,
 'B22': 45,
 'B26': 46,
 'C85': 47,
 'E17': 48,
 'B71': 49,
 'B20': 50,
 'A34': 51,
 'C86': 52,
 'A16': 53,
 'A20': 54,
 'A18': 55,
 'C54': 56,
 'C45': 57,
 'D20': 58,
 'A29': 59,
 'C95': 60,
 'E25': 61,
 'C111': 62,
 'C23 C25 C27': 63,
 'E36': 64,
 'D34': 65,
 'D40': 66,
 'B39': 67,
 'B41': 68,
 'B102': 69,
 'C123': 70,
 'E63': 71,
 'C130': 72,
 'B86': 73,
 'C92': 74,
 'A5': 75,
 'C51': 76,
 'B42': 77,
 'C91': 78,
 'C125': 79,
 'D10 D12': 80,
 'B82 

In [210]:
cabinReduced_dict.update({'n':0, 'B':5})
cabinReduced_dict

{'B': 5, 'C': 1, 'E': 2, 'D': 3, 'A': 4, 'n': 0, 'T': 6, 'F': 7, 'G': 8}

In [232]:
X_test["cabin_map"] = X_test["cabin"].map(cabin_dict)
X_train["cabin_map"] = X_train["cabin"].map(cabin_dict)
X_test["cabinReduced_map"] = X_test["cabinReduced"].map(cabinReduced_dict)
X_train["cabinReduced_map"] = X_train["cabinReduced"].map(cabinReduced_dict)
X_train.head(50)

cabin cabinReduced     sex  cabin_map  cabinReduced_map  sex_map
501       NaN            n  female          0                 0        0
588       NaN            n  female          0                 0        0
402       NaN            n  female          0                 0        0
1193      NaN            n    male          0                 0        1
686       NaN            n  female          0                 0        0
971       NaN            n    male          0                 0        1
117       E36            E  female         64                 2        0
540       NaN            n  female          0                 0        0
294       C68            C    male        157                 1        1
261       E24            E    male         30                 2        1
587       NaN            n    male          0                 0        1
489       NaN            n  female          0                 0        0
2     C22 C26            C  female          1                 1        0
405       NaN            n    male          0                 0        1
1284      NaN            n    male          0                 0        1
338       NaN            n    male          0                 0        1
356       NaN            n    male          0                 0        1
985       NaN            n    male          0                 0        1
182       NaN            n  female          0                 0        0
1027      NaN            n    male          0                 0        1
1023      NaN            n  female          0                 0        0
657       NaN            n  female          0                 0        0
891       NaN            n    male          0                 0        1
716       NaN            n    male          0                 0        1
1142      NaN            n    male          0                 0        1
520       D38            D    male        176                 3        1
1267      NaN            n  female          0                 0        0
395       NaN            n  female          0                 0        0
905       NaN            n    male          0                 0        1
278       B50            B    male        149                 5        1
239       A24            A    male        134                 4        1
1254      NaN            n    male          0                 0        1
1084      NaN            n    male          0                 0        1
110      C111            C    male         62                 1        1
1255      NaN            n    male          0                 0        1
1306      NaN            n    male          0                 0        1
822       NaN            n    male          0                 0        1
1114      NaN            n    male          0                 0        1
1110      NaN            n    male          0                 0        1
1048      NaN            n  female          0                 0        0
1034      NaN            n    male          0                 0        1
568         F            F    male        177                 7        1
615       NaN            n    male          0                 0        1
840       NaN            n  female          0                 0        0
367       NaN            n    male          0                 0        1
655       NaN            n    male          0                 0        1
1295      NaN            n    male          0                 0        1
384       NaN            n    male          0                 0        1
200        C6            C    male         12                 1        1
279       C87            C    male        150                 1        1

<h6>
    
    Zastąpienie braków wartości liczbą 0 będzie dobrym wyjściem, gdy ta liczba nie występowała wcześniej w danym zbiorze danych. W innym wypadku możemy nadać błędne wartości zmiennym i tym samym zniekształcić model czy statystyki.
    W tym wypadku liczba 0 została już przypisana innej wartości, stąd samo zastąpienie braków danych liczbą 0 nie jest dobrym pomysłem - należy zamienić wartości dla NaN oraz klucza dla którego wartość jest równa 0

<h3>Porównaj ile unikalnych wartości jest w zbiorze treningowym, a ile w
zbiorze testowym (funkcja len).
Jaka jest różnica pomiędzy liczbą etykiet przed i po redukcji oraz
mapowaniu?
Czy cały proces, który został do tej pory wykonany może mieć wpływ na
końcowy wynik predykcji i jakość modelu?

In [233]:
print('Liczba etykiet zmiennej cabinReduced test: {}'.format(len(X_test["cabinReduced"].unique())))

Liczba etykiet zmiennej cabinReduced test: 8


In [234]:
print('Liczba etykiet zmiennej cabinReduced train: {}'.format(len(X_train["cabinReduced"].unique())))

Liczba etykiet zmiennej cabinReduced train: 9


In [235]:
print('Liczba etykiet zmiennej cabinReduced_map test: {}'.format(len(X_test["cabinReduced_map"].unique())))

Liczba etykiet zmiennej cabinReduced_map test: 8


In [236]:
print('Liczba etykiet zmiennej cabinReduced_map train: {}'.format(len(X_train["cabinReduced_map"].unique())))

Liczba etykiet zmiennej cabinReduced_map train: 9


In [237]:
print('Liczba etykiet zmiennej cabin test: {}'.format(len(X_test["cabin"].unique())))

Liczba etykiet zmiennej cabin test: 71


In [238]:
print('Liczba etykiet zmiennej cabin train: {}'.format(len(X_train["cabin"].unique())))

Liczba etykiet zmiennej cabin train: 151


In [239]:
print('Liczba etykiet zmiennej cabin_map test: {}'.format(len(X_test["cabin_map"].unique())))

Liczba etykiet zmiennej cabin_map test: 71


In [240]:
print('Liczba etykiet zmiennej cabin_map train: {}'.format(len(X_train["cabin_map"].unique())))

Liczba etykiet zmiennej cabin_map train: 151


In [241]:
print('Liczba etykiet zmiennej sex test: {}'.format(len(X_test["sex"].unique())))

Liczba etykiet zmiennej sex test: 2


In [242]:
print('Liczba etykiet zmiennej sex train: {}'.format(len(X_train["sex"].unique())))

Liczba etykiet zmiennej sex train: 2


In [243]:
print('Liczba etykiet zmiennej sex_map test: {}'.format(len(X_test["sex_map"].unique())))

Liczba etykiet zmiennej sex_map test: 2


In [244]:
print('Liczba etykiet zmiennej sex_map train: {}'.format(len(X_train["sex_map"].unique())))

Liczba etykiet zmiennej sex_map train: 2


<h6>W zbiorze trenignowym jest więcej unikalnych wartości niż w zbiorze testowym (zbiór treningowy jest znacznie większy niż zbiór testowy). Redukcja etykiet w zmiennej cabin znacznie zmniejszyła ich ilość, ze 71 do 8 bądź z 151 do 9 w zależności od zbioru. Cały proces może wpłynąć na końcowy model - redukcja zmiennych i zastąpienie wartości kategorycznych wartościami liczbowymi upraszcza model, pozwala unikać zjawiska overfittingu. 

In [245]:
df.to_csv('titanic_lab3.csv')
df.head(20)
X_test.to_csv('X_test.csv')
X_train.to_csv('X_train.csv')
y_test.to_csv('y_test.csv')
y_train.to_csv('y_train.csv')